draft-ietf-privacypass-batched-tokens.md

```diff
@@ -317,15 +326,146 @@ struct {
+
+## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}
+
+The Client first creates all TokenRequest it wants to batch. To do so, the client follows
+protocol describing issuance, such as {{RFC9578, Section 5.1}} or {{RFC9578, Section 6.1}}.
+
+The Client then creates a BatchedTokenRequest structured as follows:
+
+~~~tls
+struct {
+  uint16_t token_type;
+  opaque token_request<0..2^16-1>;
+} TokenRequest
+
+struct {
+  TokenRequest token_requests<0..2^16-1>;
+} BatchTokenRequest
+~~~
+
+The structure fields are defined as follows:
+
+- "token_type" is a 2-octet integer, which matches the type of the TokenRequests in the batch.
+
+- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests.
+
+TokenRequest MUST be prefixed with a uint16 "token_type" indicating the token type. The rest of the structure follows based on that type, within the inner opaque token_request attribute.
+For instance, the TokenRequest defined in {{RFC9578}} would correspond to the following
+
+~~~tls
+struct {
+   uint16_t token_type;
+   select (token_type) {
+      case (0x0001): /* Type VOPRF(P-384, SHA-384), RFC 9578 */
+         uint8_t truncated_token_key_id;
+         uint8_t blinded_msg[Ne];
+      case (0x0002): /* Type Blind RSA (2048-bit), RFC 9578 */
+         uint8_t truncated_token_key_id;
+         uint8_t blinded_msg[Nk];
+   }
+} TokenRequest;
+~~~
+
+The Client then generates an HTTP POST request to send to the Issuer Request
+URL, with the BatchTokenRequest as the content. The media type for this request
+is "application/private-token-batch-request". An example request for the Issuer
+Request URL "https://issuer.example.net/request" is shown below.
+
+~~~
+POST /request HTTP/1.1
+Host: issuer.example.net
+Accept: application/private-token-batch-response
+Content-Type: application/private-token-batch-request
+Content-Length: <Length of BatchTokenRequest>
+
+<Bytes containing the BatchTokenRequest>
+~~~
+
+## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}
+
+Upon receipt of the request, the Issuer validates the following conditions:
+
+- The Content-Type is application/private-token-batch-request as registered with IANA.
+
+If this condition is not met, the Issuer MUST return an HTTP 400 error
+to the client.
+
+The Issuer then tries to deserialize the first 2 bytes of the i-th element
+of BatchTokenRequest.token_requests.
+If this is not a token type registered with IANA, the Issuer MUST return an HTTP 400 error
+to the client.
+The issuer creates a BatchTokenResponse structured as follows:
+
+~~~tls
+struct {
+  TokenResponse token_response; /* Defined by token_type */
```

Comment:
This isn't right. We need to have a length prefix to indicate if the field actually exists or not.

Reply:
updated. the text was correct but convoluted, I've updated it as well